### PR TITLE
Feature-bundle Enhancement (& Bug fix)

### PR DESCRIPTION
**Pull request by dhakehurst**
1) Have enhanced the feature-bundle inclusion to enable the p2 generation to also generate the feature bundles
(this is more useful than simply enabling feature-bundles to be included in the repo - though this is still possible)

2) fix a bug regarding inclusion of pre built feature-bundles

3) enabled source versions of feature-bundles to be created

### CODE REVIEW

**Comment by tombujok** [comment]
Hi David. Thank you for your contribution! Is there a way we could get in touch to discuss your changes?

**Comment by tombujok** [comment]
Hi David, I do not get the difference between a.) and b.) Could you explain it more thoroughly in the README.md?

a.) Have enhanced the feature-bundle inclusion to enable the p2 generation to also generate the feature bundles
b.) (this is more useful than simply enabling feature-bundles to be included in the repo - though this is still possible)

Why are the feature-bundles usefull? I've never used them in an Tycho build/project.

**Comment by dhakehurst** [comment]
a.) Have enhanced the feature-bundle inclusion to enable the p2 generation to also generate the feature bundles
b.) (this is more useful than simply enabling feature-bundles to be included in the repo - though this is still possible)

Difference:
  (a)  this plugin creates the feature bundle and includes it in the generate p2 bundle.
  (b) some other mechanism is used to create the feature bundle as a jar/maven artefact, this plugin includes it in the generated p2 repo.


Why are the feature-bundles usefull? I've never used them in an Tycho build/project.

When using an update site from eclipse it is better to bundles into features (and ideally categories) rather than a list of raw plugins.
 (1) got nothing to do with Tycho...in fact its about not using Tycho and using this p2-maven-plugin to generate update sites instead. I want features in that update site.
 (2) A tycho build references a p2 site generated with this p2-maven-plugin and includes features from the generated p2 site (and has feature dependencies on them) to expose to consumers of the end update site.

**Comment by dhakehurst** [comment]
made requested edits

**Comment by dhakehurst** [comment]
hmm...can't see how to merge your P2Mojo with mine?
any suggestions.
Problem area around processArtifacts().

**Comment by tombujok** [comment]
Some other pull requests have been merged, so it's will need a bit of rework...
@lukejackson Would you like to review this pull request and help @dhakehurst to merge it?
(I am looking for some active contributors ;)

**Comment by lukejackson** [comment]
@dhakehurst happy to help with merging. I'll take a look at your code and either give you some pointers on merging. Or if the merging is simple I can possibly do the merge as a pull request into your repo.

I'm not an Eclipse expert (I got involved in this project as a build engineer for some Eclipse RCP apps), so I'm not in the best position to feedback on the overall approach, but can certainly also review the code itself.

File: src/main/java/org/reficio/p2/FeatureBuilder.java [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+import static org.twdata.maven.mojoexecutor.MojoExecutor.artifactId;
+import static org.twdata.maven.mojoexecutor.MojoExecutor.configuration;
+import static org.twdata.maven.mojoexecutor.MojoExecutor.element;
+import static org.twdata.maven.mojoexecutor.MojoExecutor.executeMojo;
+import static org.twdata.maven.mojoexecutor.MojoExecutor.executionEnvironment;
+import static org.twdata.maven.mojoexecutor.MojoExecutor.goal;
+import static org.twdata.maven.mojoexecutor.MojoExecutor.groupId;
+import static org.twdata.maven.mojoexecutor.MojoExecutor.name;
+import static org.twdata.maven.mojoexecutor.MojoExecutor.plugin;
+import static org.twdata.maven.mojoexecutor.MojoExecutor.version;
```
Comment (lukejackson) [inline src/main/java/org/reficio/p2/FeatureBuilder.java]: these imports are not used

File: src/main/java/org/reficio/p2/FeatureBuilder.java [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+import org.apache.commons.io.FileUtils;
+import org.apache.commons.io.IOUtils;
+import org.reficio.p2.bundler.ArtifactBundlerInstructions;
+import org.reficio.p2.utils.Utils;
+import org.reficio.p2.utils.XmlUtils;
+import org.w3c.dom.Document;
+import org.w3c.dom.Element;
+import org.w3c.dom.NodeList;
+
+import com.google.common.collect.Multimap;
```
Comment (lukejackson) [inline src/main/java/org/reficio/p2/FeatureBuilder.java]: not used

File: src/main/java/org/reficio/p2/FeatureBuilder.java [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+
+import aQute.lib.osgi.Jar;
```
Comment (lukejackson) [inline src/main/java/org/reficio/p2/FeatureBuilder.java]: not used

File: src/main/java/org/reficio/p2/FeatureBuilder.java [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+
+public class FeatureBuilder {
+
+	public FeatureBuilder(P2FeatureDefinition p2FeatureDefintion, Map<P2Artifact, ArtifactBundlerInstructions>  bundlerInstructions) {
+		this.p2FeatureDefintion = p2FeatureDefintion;
+		this.bundlerInstructions = bundlerInstructions;
+	}
+
+	Map<P2Artifact, ArtifactBundlerInstructions>  bundlerInstructions;
+	P2FeatureDefinition p2FeatureDefintion;
```
Comment (lukejackson) [inline src/main/java/org/reficio/p2/FeatureBuilder.java]: i find it clearer to declare all your fields at the beginning of the class, and make them private and final wherever possible

File: src/main/java/org/reficio/p2/FeatureBuilder.java [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+			licenceElement.setTextContent(this.p2FeatureDefintion.getLicense());
+		}
+
+		for(P2Artifact artifact: this.p2FeatureDefintion.artifacts) {
+			Element pluginElement = XmlUtils.createElement(xmlDoc,featureElement,"plugin");
+			String id = this.bundlerInstructions.get(artifact).getSymbolicName();
+			String version = this.bundlerInstructions.get(artifact).getProposedVersion();
+			pluginElement.setAttribute("id", id);
+			pluginElement.setAttribute("download-size", "0"); //TODO
+			pluginElement.setAttribute("install-size", "0");  //TODO
```
Comment (lukejackson) [inline src/main/java/org/reficio/p2/FeatureBuilder.java]: is this OK to set 0?

File: src/main/java/org/reficio/p2/FeatureBuilder.java [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+			JarOutputStream jar = new JarOutputStream(fos, mf);
+			addToJar(jar, sourceFeatureContent);
+		} catch (Exception e) {
+			throw new RuntimeException("Cannot generate feature", e);
+		}
+	}
+
+
+	void createFeatureWithTycho() {
+	}
```
Comment (lukejackson) [inline src/main/java/org/reficio/p2/FeatureBuilder.java]: unused - best to delete

File: src/main/java/org/reficio/p2/FeatureBuilder.java [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+			pluginElement.setAttribute("unpack", "false");
+			
+		}
+		
+		try {
+			File sourceFeatureContent = new File(destinationFolder, this.getFeatureFullName()+".source");
+			sourceFeatureContent.mkdir();
+			XmlUtils.writeXml(this.xmlDoc, new File(sourceFeatureContent, "feature.xml"));
+			
+			//TODO: add other files that are required by the feature
```
Comment (lukejackson) [inline src/main/java/org/reficio/p2/FeatureBuilder.java]: does this need to be done before merging?

File: src/main/java/org/reficio/p2/TychoFeatureBuilder.java [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+import org.reficio.p2.publisher.BundlePublisher.Builder;
+import org.twdata.maven.mojoexecutor.MojoExecutor;
+
+public class TychoFeatureBuilder {
+    private static final String TYCHO_VERSION = "0.18.1";
+
+    File featureFile;
+    private final String outputDirectory;
+    String featureId;
+    String featureVersion;
```
Comment (lukejackson) [inline src/main/java/org/reficio/p2/TychoFeatureBuilder.java]: can these be private and final?

File: src/main/java/org/reficio/p2/utils/JarUtils.java [hunk 9]
```diff
@@ -25,16 +25,31 @@
...
+import org.apache.commons.io.filefilter.FileFilterUtils;
+import org.apache.commons.io.filefilter.RegexFileFilter;
+import org.apache.commons.io.filefilter.WildcardFileFilter;
+import org.apache.maven.plugin.logging.Log;
 import org.w3c.dom.Document;
+import org.w3c.dom.Element;
+import org.w3c.dom.Node;
+import org.w3c.dom.NodeList;
+
+import clover.retrotranslator.edu.emory.mathcs.backport.java.util.Arrays;
```
Comment (lukejackson) [inline src/main/java/org/reficio/p2/utils/JarUtils.java]: bogus import?

File: src/main/java/org/reficio/p2/utils/JarUtils.java [hunk 10]
```diff
@@ -103,59 +118,59 @@ public static void adjustFeatureQualifierVersionWithTimestamp(File inputFile, Fi
...
+    public static void adjustFeaturePluginData(Document featureSpec, File pluginDir, Log log) throws IOException {
+	        //get list of all plugins
+	        NodeList plugins = featureSpec.getElementsByTagName("plugin");
+	        for(int i=0; i<plugins.getLength(); ++i) {
+	        	Node n = plugins.item(i);
+	        	if (n instanceof Element) {
+		        	Element el = (Element)n;
+		        	String pluginId = el.getAttribute("id");
+		        	File[] files = findFiles(pluginDir, pluginId);
+		        	if (files.length < 0) {
```
Comment (lukejackson) [inline src/main/java/org/reficio/p2/utils/JarUtils.java]: always false. i assume you mean == 0?

**Comment by lukejackson** [comment]
The merge of P2Mojo.java is a bit tricky, without knowing the full context of your change. 

The external change that has been made to this class since you forked is relatively straightforward however. Essentially processArtifacts has been split into two stages - processRootArtifacts() which processes the artifacts directly referenced in the plugin config, and processTransitiveArtifacts() which processes the transitive artifacts of the root artifacts. Additionally a check has been added to prevent the same artifact being bundled more than once. Preventing duplicate bundling is a simple optimisation, and also prevents an artifact which happens to be a root artifact and a transitive artifact of another root artifact being bundled twice, once with instructions, and once without instructions.

You will need to reapply your change to processArtifacts() to both processRootArtifacts() and processTransitiveArtifacts().

**Comment by tombujok** [comment]
Thanks for the contribution. I am closing this one since it's been merged here #75

**Pull request closed without merging**